### ISSUE TITLE
STRING with more than one word is not treated properly as ELEMENT with the motion commands 

### ISSUE THREAD

**Issue by nimaai** [guns/vim-sexp#33]
```
(foo "bar" :baz)
 ^   ^     ^
```
starting at `f` jumps properly with `<M-w>`

```
(foo "bar qux" :baz)
 ^   ^    ^    ^
```
jumps **inside** the string with `<M-w>`

same behaviour with `<M-b>`

from my point of view this is an incorrect behaviour and it was not working like this in the past.

**Comment by bpstahlman** [comment]
@nimaai Thanks for reporting. I've confirmed the issue, and will post an update once I've had a chance to root-cause...

**Comment by nimaai** [comment]
and also (not sure if this is related) when pressing `"` inside the string at the end i would expect this plugin to just "jump over" in order to maintain the balanced closing `"`. 

correct:
```
"foo bar"
       ^
 -->
 "foo bar"
          ^
```

but what happens is following:
```
"foo bar"
       ^
 -->
 "foo bar "" "
          ^
```

**Comment by bpstahlman** [comment]
@nimaai Interestingly, I've been able to reproduce the original issue only in the terminal. Have you tried this in gvim or only in a terminal?

**Comment by nimaai** [comment]
@bpstahlman i am using terminal neovim and gui neovide. same issues in both environments.

**Comment by bpstahlman** [comment]
@nimaai Which version of neovim? I see the original problem you reported in Vim 8.1, but not in Vim 8.2, so I want to be sure I'm using the same version of neovim.

**Comment by bpstahlman** [comment]
Also, can you tell me the version/commit of both vim-sexp and whatever lisp syntax file you're using? The latter may be significant because the only time I was able to reproduce the issue, it occurred because the string wasn't being highlighted properly, and vim-sexp uses the synstack to delineate the elements.

I don't see the issue on neovim 0.6.1. So far, I've seen it only on Vim 8.1.

**Comment by nimaai** [comment]
@bpstahlman 
```
neovim v0.10.2
vim-sexp 14464d4
clojure.vim 128ccd4
```
same problem with `vim-racket dcfb8c1`.

**Comment by bpstahlman** [comment]
I'm not able to reproduce either issue with the versions you mentioned. The only other variable that seems potentially relevant is the terminal itself. I would expect this to matter more with Vim than Neovim. For instance, in my current terminal, Vim will treat a `<M-w>` like a simple `w` if I don't have a `set <M-w>=^[w` in my vimrc. IIUC, Neovim handles that sort of thing transparently, with no need for special termcap sets. But just in case, which terminal are you using?

**Comment by nimaai** [comment]
@bpstahlman iterm Build 3.5.10

**Comment by hoclun-rigsep** [comment]
Just noticed this issue myself recently. Kitty term recent version, neovim 0.10.2

**Comment by bpstahlman** [comment]
> setting `additional_vim_regex_highlighting = true` in treesitter config solved this issue for me

@AlexKehl Great! Thanks for looking into this. I'll try to add a comment to the README. Incidentally, I'm hoping to merge the following branch within the next few days:

([enhanced-sexp-objects](https://github.com/bpstahlman/vim-sexp/tree/enhanced-sexp-objects))

It adds significant new functionality: e.g., a new set of sexp traversal commands known as "flow" commands, clone commands, enhanced indenting and whitespace cleanup, etc... Would greatly appreciate feedback from active users who have an easy way to switch to the fork to try it out...

**Comment by bpstahlman** [comment]
@AlexKehl @nimaai @hoclun-rigsep 
Created a PR (https://github.com/guns/vim-sexp/pull/34) with a summary of new features...

**Comment by bpstahlman** [comment]
I was planning to merge it within the next few days, but would feel better
if someone besides me had a chance to try it out first. This issue
highlights the advantage of testing with multiple setups…

**Comment by nimaai** [comment]
@bpstahlman i am using your PR branch but the reported problems are not fixed there neither. 🙏

**Comment by bpstahlman** [comment]
> [@bpstahlman](https://github.com/bpstahlman) i am using your PR branch but the reported problems are not fixed there... 🙏

Are  you using treesitter?
Did you try setting `additional_vim_regex_highlighting = true`, as suggested by @AlexKehl ?

**Comment by nimaai** [comment]
oh, i had this setting top level in the treesitter config and not under `highlight`. now both issues are resolved. yay

**Comment by bpstahlman** [comment]
> oh, i had this setting top level in the treesitter config and not under `highlight`. now both issues are resolved. yay

Ah, yes. Nested option tables are flexible, but also easy to break...
Appreciate your taking the time to try it out.

**Comment by AlexKehl** [comment]
Sadly setting `additional_vim_regex_highlighting = true` slows nvim down (as described [here](https://github.com/nvim-treesitter/nvim-treesitter?tab=readme-ov-file#modules))
Are there actually any plans in utilizing treesitter in `vim-sexp` in the future?

**Comment by bpstahlman** [comment]
Yes. I think this needs to be done. I noticed while debugging a different
issue that the logic for ignoring brackets in comments assumes
non-treesitter syntax regions. This sort of check is cleaner with
treesitter. The plugin should use it if available, falling back to syntax
regions. I'll add an issue for it…

**Comment by bpstahlman** [comment]
It also bothers me that sexp's motions land on Treesitter "extra" nodes (e.g., comments).
This behavior is less problematic in non-Treesitter buffers, where at least the whole comment is treated as a single element. In a Treesitter buffer, by contrast, each word is effectively a separate element.

I welcome suggestions on how vim-sexp should treat "extra" and unnamed nodes when Treesitter support is added...

**Comment by bpstahlman** [comment]
@nimaai @AlexKehl @hoclun-rigsep 
I've just pushed another commit to the PR branch, which fixes several issues: e.g., animation with smear-cursor and incorrect handling of counts with certain commands.

I'm planning to start looking into the addition of Neovim-specific Treesitter logic to fix handling of comments and strings...

**Comment by nimaai** [comment]
@bpstahlman i have found an issue with your branch when indenting a visual selection:

```
Error detected while processing function sexp#indent:
line   74:
E493: Backwards range given: 22,0call synID(line("."), col("."), 1)
```
`master` on `guns/vim-sexp` is find.

**Comment by bpstahlman** [comment]
@nimaai I'm able to reproduce it now and will post an update as soon as a fix is available.
I had trouble reproducing it at first because I was using characterwise-visual mode selections, and the issue seems to occur mostly in linewise-visual mode.

**Comment by bpstahlman** [comment]
@nimaai I believe this should fix the issue.

**Comment by nimaai** [comment]
@bpstahlman yes, it's alright now. thanks

**Comment by bpstahlman** [comment]
@nimaai @AlexKehl @hoclun-rigsep 
Excellent! Thanks for testing. Incidentally, work on incorporating treesitter logic to fix the original issue is well underway. Will post once it's complete and pushed somewhere (maybe the PR branch, maybe another). With these changes, it should no longer be necessary to set `additional_vim_regex_highlighting`.

**Comment by hoclun-rigsep** [comment]
branch = "ws-cleanup-refactor"

Just let me know if this is the branch you want me on right now!

**Comment by bpstahlman** [comment]
> ```
> 	branch = "ws-cleanup-refactor"
> ```
> 
> Just let me know if this is the branch you want me on right now!

At this point, it would be best to switch to the PR branch ([enhanced-sexp-objects](https://github.com/bpstahlman/vim-sexp/tree/enhanced-sexp-objects)), which includes everything on ws-cleanup-refactor, and is the branch slated for merge to master.

**Comment by bpstahlman** [comment]
> [@bpstahlman](https://github.com/bpstahlman)
> 
> > At this point, it would be best to switch to the PR branch
> 
> i am on that branch since some time now. i will report back if i find something

Yes, but a significant amount of new functionality was merged into the PR branch only yesterday, so you should make sure your plugin manager pulls the latest.
The PR comment has been updated to reflect the added functionality.
In addition to Treesitter support and trailing comment alignment, there have been a few significant bugfixes.

**Comment by bpstahlman** [comment]
Several more fixes/enhancements have been added to the branch within the past few days.
Unless a showstopper issue is reported before then, I'm planning to merge PR #34 this weekend.
